agent/engine/docker_task_engine_test.go

```diff
@@ -1513,3 +1513,91 @@ func TestCreateContainerOnAgentRestart(t *testing.T) {
+	accessKeyID := "akid"
+	secretAccessKey := "sakid"
+	sessionToken := "token"
+	executionRoleCredentials := credentials.IAMRoleCredentials{
+		CredentialsID:   credentialsID,
+		AccessKeyID:     accessKeyID,
+		SecretAccessKey: secretAccessKey,
+		SessionToken:    sessionToken,
+	}
+
+	testTask := testdata.LoadTask("sleep5")
+	// Configure the task and container to use execution role
+	testTask.SetExecutionRoleCredentialsID(credentialsID)
+	testTask.Containers[0].RegistryAuthentication = &api.RegistryAuthenticationData{
+		Type: "ecr",
+		ECRAuthData: &api.ECRAuthData{
+			UseExecutionRole: true,
+		},
+	}
+	container := testTask.Containers[0]
+
+	credentialsManager.EXPECT().GetTaskCredentials(credentialsID).Return(credentials.TaskIAMRoleCredentials{
+		ARN:                "",
+		IAMRoleCredentials: executionRoleCredentials,
+	}, true)
+
+	client.EXPECT().PullImage(gomock.Any(), gomock.Any()).Do(
+		func(image string, auth *api.RegistryAuthenticationData) {
+			assert.Equal(t, container.Image, image)
+			assert.Equal(t, auth.ECRAuthData.GetPullCredentials(), executionRoleCredentials)
+		}).Return(DockerContainerMetadata{})
+	imageManager.EXPECT().RecordContainerReference(container).Return(nil)
+	imageManager.EXPECT().GetImageStateFromImageName(container.Image)
+
+	taskEngine.(*DockerTaskEngine).pullContainer(testTask, container)
+}
+
+// TestTasktionRoleOnRestart tests the agent will process the task recorded in
+// the state file on restart
+func TestNewTaskTransitionOnRestart(t *testing.T) {
+	ctrl, _, _, taskEngine, _, _ := mocks(t, &defaultConfig)
+	defer ctrl.Finish()
+
+	dockerTaskEngine := taskEngine.(*DockerTaskEngine)
+	state := dockerTaskEngine.State()
+
+	testTask := testdata.LoadTask("sleep5")
+	// add the task to the state to simulate the agent restored the state on restart
+	state.AddTask(testTask)
+	// Set the task to be stopped so that the process can done quickly
+	testTask.SetDesiredStatus(api.TaskStopped)
+
+	dockerTaskEngine.synchronizeState()
+	_, ok := dockerTaskEngine.managedTasks[testTask.Arn]
+	assert.True(t, ok, "task wasnot started")
+}
+
+// TestNotTranistionTaskUseExectionRole tests on agent restart task requires execution
+// role credentials shouldnot be started
+func TestNotTranistionTaskUseExectionRole(t *testing.T) {
+	ctrl, _, _, taskEngine, _, _ := mocks(t, &defaultConfig)
+	defer ctrl.Finish()
+
+	dockerTaskEngine := taskEngine.(*DockerTaskEngine)
+	state := dockerTaskEngine.State()
+
+	testTask := testdata.LoadTask("sleep5")
+	// add the task to the state to simulate the agent restored the state on restart
+	state.AddTask(testTask)
+	testTask.Containers[0].RegistryAuthentication = &api.RegistryAuthenticationData{
+		Type: "ecr",
+		ECRAuthData: &api.ECRAuthData{
+			UseExecutionRole: true,
+		},
+	}
+
+	dockerTaskEngine.synchronizeState()
+	_, ok := dockerTaskEngine.managedTasks[testTask.Arn]
+	assert.False(t, ok, "task require credentials shouldnot be started on agent restart")
```

Comment:
I'm not sure what this is saying, but maybe one of these?
The task requires credentials, so it should not be started on agent restart.
or
Tasks that require credentials should not be started on agent restart.